File: typescript/ocr-numbers/ocr-numbers.ts
class DefaultParser {
    value: string = '?'
}

class Parser {
    private pattern: string[][]
    value: string

    constructor(_value: string, _pattern: string) {
        this.value = _value
        this.pattern = _pattern
            .split('\n')
            .map(line => line.split(''))
    }

    canParse(_line: number, _position: number, text: string[][]): boolean {
        return this.pattern.every((line, lIdx) => {
            return line.every((character, cIdx) => {
                let l = (_line) * 4 + lIdx
                let p = (_position) * 3 + cIdx
                return character == text[l][p]
            })
        })
    }
}

export default class OcrParser {
    private static parsers: Parser[] = [
        new Parser('0', " _ \n| |\n|_|"),
        new Parser('1', "   \n  |\n  |"),
        new Parser('2', " _ \n _|\n|_"),
        new Parser('3', " _ \n _|\n _|"),
        new Parser('4', "   \n|_|\n  |"),
        new Parser('5', " _ \n|_ \n _|"),
        new Parser('6', " _ \n|_ \n|_|"),
        new Parser('7', " _ \n  |\n  |"),
        new Parser('8', " _ \n|_|\n|_|"),
        new Parser('9', " _ \n|_|\n _|")

    ]
    static convert(text: string): string {
        let characters = text.split('\n')
            .map(c => c.split(''))

        let length = characters[0].length / 3
        let lines = characters.length / 4

        let result = ''
        for (let l = 0; l < lines; l++) {
            if (l > 0) {
                result += ','
            }

            for (let i = 0; i < length; i++) {
                let parser = OcrParser.parsers.find(p => p.canParse(l, i, characters)) || new DefaultParser()
                result += parser.value
            }
        }

        return result
    }
}

File: scala/ocr-numbers/src/main/scala/OcrNumbers.scala
object OcrNumbers {
  private val lineWidth = 3
  private val lineHeight = 4

  def convert(text: Seq[String]): String = {
    val isValid = text.length % lineHeight == 0 && text.forall(_.length % lineWidth == 0)

    if (!isValid) "?"
    else {
      val characters = text.grouped(lineHeight).map(chunk)
      characters.map { line => line.map { case Parser(v) => v }.mkString("") }.mkString(",")
    }
  }

  private def chunk(line: Seq[String]): Seq[Seq[String]] = {
    val fragments = line
      .map(_.grouped(lineWidth).toIndexedSeq)
      .toIndexedSeq

    for { i <- fragments.head.indices } yield Seq(fragments(0)(i), fragments(1)(i), fragments(2)(i), fragments(3)(i))
  }
}

sealed trait Parser {
  val pattern: Seq[String]
  val value: String
  def parse(text: Seq[String]): Option[String] = {
    if (pattern.zip(text).forall(pair => pair._1 == pair._2)) Some(value)
    else None
  }

  def unapply(arg: Seq[String]): Option[String] = this.parse(arg)
}

object Parser {
  def unapply(arg: Seq[String]): Option[String] = {
    val result = arg match {
      case ZeroParser(v) => v
      case OneParser(v) => v
      case TwoParser(v) => v
      case ThreeParser(v) => v
      case FourParser(v) => v
      case FiveParser(v) => v
      case SixParser(v) => v
      case SevenParser(v) => v
      case EightParser(v) => v
      case NineParser(v) => v
      case _ => "?"

    }

    Some(result)
  }
}

object ZeroParser extends Parser {
  override val pattern: Seq[String] = Seq(
    " _ ",
    "| |",
    "|_|",
    "   ")

  override val value: String = "0"
}

object OneParser extends Parser {
  override val pattern: Seq[String] = Seq(
    "   ",
    "  |",
    "  |",
    "   ")

  override val value: String = "1"
}

object TwoParser extends Parser {
  override val pattern: Seq[String] = Seq(
    " _ ",
    " _|",
    "|_ ",
    "   ")

  override val value: String = "2"
}

object ThreeParser extends Parser {
  override val pattern: Seq[String] = Seq(
    " _ ",
    " _|",
    " _|",
    "   ")

  override val value: String = "3"
}

object FourParser extends Parser {
  override val pattern: Seq[String] = Seq(
    "   ",
    "|_|",
    "  |",
    "   ")

  override val value: String = "4"
}

object FiveParser extends Parser {
  override val pattern: Seq[String] = Seq(
    " _ ",
    "|_ ",
    " _|",
    "   ")

  override val value: String = "5"
}

object SixParser extends Parser {
  override val pattern: Seq[String] = Seq(
    " _ ",
    "|_ ",
    "|_|",
    "   ")

  override val value: String = "6"
}

object SevenParser extends Parser {
  override val pattern: Seq[String] = Seq(
    " _ ",
    "  |",
    "  |",
    "   ")

  override val value: String = "7"
}

object EightParser extends Parser {
  override val pattern: Seq[String] = Seq(
    " _ ",
    "|_|",
    "|_|",
    "   ")

  override val value: String = "8"
}

object NineParser extends Parser {
  override val pattern: Seq[String] = Seq(
    " _ ",
    "|_|",
    " _|",
    "   ")

  override val value: String = "9"
}